src/docfx/restore/RestoreLocker.cs

```diff
@@ -0,0 +1,89 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Diagnostics;
+using System.IO;
+using System.Threading.Tasks;
+
+namespace Microsoft.Docs.Build
+{
+    internal class RestoreLocker
+    {
+        private readonly RestoreLock _restoreItem;
+
+        public RestoreLocker(string docset)
+        {
+            Debug.Assert(!string.IsNullOrEmpty(docset));
+
+            _restoreItem = Load(docset).Result;
+        }
+
+        public bool TryGetWorkTreeHead(string href, out string workTreeHead)
+        {
+            return _restoreItem.Git.TryGetValue(href, out workTreeHead);
+        }
+
+        public static Task Lock(string docset, Func<RestoreLock, RestoreLock> process)
+        {
+            Debug.Assert(!string.IsNullOrEmpty(docset));
+
+            var restoreLockFilePath = GetRestoreLockFilePath(docset);
+            return ProcessUtility.ProcessLock(
+                Path.GetRelativePath(AppData.RestoreLockDir, restoreLockFilePath),
+                () =>
+                {
+                    Directory.CreateDirectory(Path.GetDirectoryName(restoreLockFilePath));
+                    using (var fileStream = File.Open(restoreLockFilePath, FileMode.OpenOrCreate, FileAccess.ReadWrite))
+                    {
+                        var sr = new StreamReader(fileStream);
+                        var sw = new StreamWriter(fileStream);
+
+                        // read restore item
+                        var (_, restoreItem) = JsonUtility.Deserialize<RestoreLock>(sr.ReadToEnd());
+                        if (restoreItem == null)
+                            restoreItem = new RestoreLock();
+
+                        // process restore item
+                        restoreItem = process(restoreItem);
+
+                        // write back restore item
+                        fileStream.SetLength(0);
+                        sw.Write(JsonUtility.Serialize(restoreItem));
+                        sw.Flush();
+
+                        return Task.CompletedTask;
+                    }
+                });
+        }
+
+        private static async Task<RestoreLock> Load(string docset)
+        {
+            Debug.Assert(!string.IsNullOrEmpty(docset));
+
+            var restoreLockFilePath = GetRestoreLockFilePath(docset);
+            var restore = new RestoreLock();
+            await ProcessUtility.ProcessLock(
+                Path.GetRelativePath(AppData.RestoreLockDir, restoreLockFilePath),
+                () =>
+                {
+                    if (File.Exists(restoreLockFilePath))
+                    {
+                        (_, restore) = JsonUtility.Deserialize<RestoreLock>(File.ReadAllText(restoreLockFilePath));
```

Comment:
>restoreLockFilePath [](start = 93, length = 19)

Can you read this file sucessfully, when it's already used as ProcessUtility's lock file?

Reply:
it's not used as ProcessUtility's lock file

---
In reply to: [200900276](https://github.com/dotnet/docfx/pull/3020#discussion_r200900276) [](ancestors = 200900276)

Reply:
Got it. They are under different directories. 

---
In reply to: [200903500](https://github.com/dotnet/docfx/pull/3020#discussion_r200903500) [](ancestors = 200903500,200900276)